### ISSUE TITLE
NVIDIA nv0 training stalls beyond 60s compiling native artifact for LightGBM predictor kernel

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#346]
Public `recipe.train().optimizer(adamw).lr(0.0001).seed(109315165187836).epochs(1).tf(32).run()` on `backend:nv0` remains active after 64s. The first block reaches the stall first:

recipe.model()
.lgbm().log().norm(batch).iq(4).xs
.bayes().elu().qi(4).0
.loss(mse);

Earliest failure is `recipe.train().run()` -> `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` on the NVIDIA branch (`ptx`/`ptxas`) for the predictor primitive produced by `fit_lightgbm` (recipe.rs:7437) -> `boosted_predictor` (7344). The `data/ordinal/ordered_subfolders` loader and second `.bayes()` block are after or ignored by the same loader; they do not change the earliest path. Measured 64s is materially slower than a small ordinal load with default LightGBM histogram trees justifies; dominant gap is the unreported NVIDIA native compilation stall for the LightGBM predictor kernel with `.iq(4).xs` integer quantization.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `6954861a73b74c29`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling-small:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "NVIDIA nv0 training stalls beyond 60s compiling native artifact for LightGBM predictor kernel",
  "body": "Public `recipe.train().optimizer(adamw).lr(0.0001).seed(109315165187836).epochs(1).tf(32).run()` on `backend:nv0` remains active after 64s. The first block reaches the stall first:\n\nrecipe.model()\n.lgbm().log().norm(batch).iq(4).xs\n.bayes().elu().qi(4).0\n.loss(mse);\n\nEarliest failure is `recipe.train().run()` -> `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` on the NVIDIA branch (`ptx`/`ptxas`) for the predictor primitive produced by `fit_lightgbm` (recipe.rs:7437) -> `boosted_predictor` (7344). The `data/ordinal/ordered_subfolders` loader and second `.bayes()` block are after or ignored by the same loader; they do not change the earliest path. Measured 64s is materially slower than a small ordinal load with default LightGBM histogram trees justifies; dominant gap is the unreported NVIDIA native compilation stall for the LightGBM predictor kernel with `.iq(4).xs` integer quantization.",
  "rationale": "Searched open nv0 performance issues (282, 271, 345, 326, 325, 324, 317, 315, 293, and others); none cover a LightGBM predictor compilation stall on nv0. Issue 282 covers svm/kmeans predictor with integer arithmetic; 277 covers CPU LightGBM predictor; both use different implementation paths (lower_estimator vs fit_lightgbm / native_cpu vs native_nvidia). The packet's `.lgbm()` reaches `fit_lightgbm` -> `boosted_predictor` -> `compile_native_artifact` (recipe.rs 7437/7344/2828) on `BackendTarget::Nvidia`, which is unrepresented. The ordinal subfolder loader is a new input representation for this path, materially expanding the unreported defect. Therefore `new` with earliest public failure `recipe.train().run()` stall during NVIDIA native artifact compilation for LightGBM predictor."
}
```

## Deterministic failure packet

kind=performance
id=6954861a73b74c29
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:10004 next:10005 composition:109315165187836
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 64 seconds
measurement=elapsed_seconds:64 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /tmp/recipe-composition-repro-nv0-10004-333.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/ordinal/ordered_subfolders")
		.target("target");
	let model = recipe.model()
		.lgbm().log().norm(batch).iq(4).xs
		.bayes().elu().qi(4).0
		.loss(mse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(109315165187836)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.tf(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
The observed stall is at `compile_native_artifact` for the LightGBM predictor on the NVIDIA backend, which is the same earliest failure path as issue #346. This packet adds further evidence for that issue.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `74b6f5d4b294a70d`
- Classifier provider: `agy`
- Classifier model: `gpt-oss-120b-medium`
- Classifier effort: `medium`

### Classifier decision

```json
{"body":"The observed stall is at `compile_native_artifact` for the LightGBM predictor on the NVIDIA backend, which is the same earliest failure path as issue #346. This packet adds further evidence for that issue.","issue":346,"rationale":"Source inspection (compile_model, compile_native_artifact, fit_lightgbm) confirms the earliest failure matches issue 346. No new implementation path is introduced, so the correct classification is a comment on the existing issue.","title":"NVIDIA nv0 training stalls during native artifact compilation for LightGBM predictor","verdict":"comment"}
```

## Deterministic failure packet

kind=performance
id=74b6f5d4b294a70d
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=clean
cursor=seed:0 cursor:13372 next:13373 composition:8055272499188
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-909-13372/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.attn(1).selu().norm(batch).iq(2).xs
		.attn(1).prelu().iq(2).xxs
		.loss(focal);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(8055272499188)
		.epochs(1)
		.log(all)
		.bf(16)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Public `recipe.train().optimizer(adamw).lr(0.001).seed(109315165187836).epochs(1).tf(32).run()` on `backend:nv0` remains active after 60 seconds. Earliest failure is `recipe.train().run()` → `compile_model` → `NativeModelIr::from_graph` → `emit` → `compile_native_artifact` on the NVIDIA branch for the LightGBM predictor (`fit_lightgbm` → `boosted_predictor`). The `data/ordinal/ordered_subfolders` loader and `.iq(4).xs` quantization do not affect the earliest path.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `572d113528adfee4`
- Classifier provider: `agy`
- Classifier model: `gpt-oss-120b-medium`
- Classifier effort: `medium`

### Classifier decision

```json
{"body":"Public `recipe.train().optimizer(adamw).lr(0.001).seed(109315165187836).epochs(1).tf(32).run()` on `backend:nv0` remains active after 60 seconds. Earliest failure is `recipe.train().run()` → `compile_model` → `NativeModelIr::from_graph` → `emit` → `compile_native_artifact` on the NVIDIA branch for the LightGBM predictor (`fit_lightgbm` → `boosted_predictor`). The `data/ordinal/ordered_subfolders` loader and `.iq(4).xs` quantization do not affect the earliest path.","issue":346,"rationale":"The packet follows the same earliest failing operation and implementation path as issue 346 (LightGBM predictor compilation stall on NVIDIA backend). The new loader and quantization variant are additional evidence expanding the existing defect, not a distinct cause, so the appropriate action is a comment on issue 346.","title":"NVIDIA nv0 training stalls beyond 60s compiling native artifact for LightGBM predictor kernel","verdict":"comment"}
```

## Deterministic failure packet

kind=performance
id=572d113528adfee4
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=clean
cursor=seed:0 cursor:12935 next:12936 composition:24259008307045
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-1007-12935/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/numeric/samples_sqlite.sqlite")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.knn(3).gelu().iq(3).s
		.cbst().relu().norm(batch).iq(2).s
		.loss(focal);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(24259008307045)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.fp(64)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #435.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #435: Verify NVIDIA LightGBM compilation after data loading</summary>

Closes #346

## Verification

- Ran the exact reported source on current minimal. It stopped before compilation in 0.07 seconds with `dataset has no training features` because the one-line class-subfolder samples currently fall through to target-only table parsing.
- Reran the unchanged source at the already-pending directory-loader head `ca69266d`. Only the saved bundle path changed from `/tmp` to `/home/nate/.cache/recipe/issues/346/bundle.ogdl` because `/tmp` is tmpfs.
- With the loader boundary corrected, the real loader produced 1,000 rows with an input width of 23 and reached the full LightGBM plus Bayes TF32 path.
- All four staged workload native artifacts were cache misses. Each completed Clang LLVM bitcode generation and `llc` PTX generation. The separate 64-row calibration artifact was a cache hit.
- The process exited 0 in 8.50 seconds. The epoch took 12.466 milliseconds, MSE was finite at 6.0000, and the saved bundle was 26,973 bytes.
- The reported NVIDIA native-artifact compile stall does not reproduce after the existing data-loading boundary is crossed.

## Change

This verification commit changes no tracked files. It does not duplicate the pending directory-loader correction, and current compiler behavior does not justify a production-code change for this issue.

</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.

**Comment by nm-z** [comment]
The public `recipe.train().run()` remains active after 60 seconds on `backend:nv0`. The earliest public failure is native artifact compilation for the first `.lgbm()` block, through `fit_lightgbm` -> `boosted_predictor` -> NVIDIA LLVM/PTX and `ptxas` compilation. This roughly 12,000-row, 1-feature workload performs about 11 million histogram-bin operations per boosting pass set, with 8 boosting iterations and at most 7 leaf expansions per tree, plus minimal tree prediction and memory traffic. The expected end-to-end runtime is under a few seconds, so the dominant gap is the native NVIDIA compiler stall rather than epoch arithmetic.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `4db372e6166e74c7`
- Classifier provider: `copilot`
- Classifier model: `gpt-5.6-luna`
- Classifier effort: `auto`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 346,
  "title": "NVIDIA nv0 training stalls beyond 60s compiling native artifact for LightGBM predictor kernel",
  "body": "The public `recipe.train().run()` remains active after 60 seconds on `backend:nv0`. The earliest public failure is native artifact compilation for the first `.lgbm()` block, through `fit_lightgbm` -> `boosted_predictor` -> NVIDIA LLVM/PTX and `ptxas` compilation. This roughly 12,000-row, 1-feature workload performs about 11 million histogram-bin operations per boosting pass set, with 8 boosting iterations and at most 7 leaf expansions per tree, plus minimal tree prediction and memory traffic. The expected end-to-end runtime is under a few seconds, so the dominant gap is the native NVIDIA compiler stall rather than epoch arithmetic.",
  "rationale": "Issue 346 and its comments cover the same earliest public failure and implementation path: `recipe.train().run()` stalls in `compile_model` while compiling the predictor generated by `fit_lightgbm` and `boosted_predictor` through the NVIDIA PTX and `ptxas` branch. The current `.lgbm()` composition reaches that path before later Bayes, activation, normalization, quantization, loss, or epoch configuration can establish another cause. The temporal chronological-splits loader is additional evidence for the same defect. The measured 60 seconds is materially slower than the small workload justifies, with expected work in the low millions of scalar operations and tens of megabytes of memory traffic, while the dominant delay is native artifact compilation."
}
```

## Deterministic failure packet

kind=performance
id=4db372e6166e74c7
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:23074 next:23075 composition:79931167432646
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-7132-23074/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.layer(4).elu().norm(batch).iq(3).xs
		.rnn(8).iq(3).xxs
		.loss(focal);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(79931167432646)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.tf(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0
